src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+
+                // Await the result and get its value.
+                await resultTask.ConfigureAwait(false);
+                var resultProperty = resultTask.GetType().GetProperty(nameof(Task<>.Result))
+                    ?? throw new InvalidOperationException($"Unexpected return type from {typeFullName}:{executeMethodName} in assembly {assemblyPath}, expected type Task<>.");
+                var result = resultProperty.GetValue(resultTask);
+
+                // CustomResponse.Message must express positions in CustomMessage.TextDocument as indexes referencing CustomResponse.Positions.
+                // LinePositionWriteConverter allows serializing extender-defined types into json with indexes referencing LinePosition objects.
+                JsonSerializerOptions writeOptions = new();
+                LinePositionWriteConverter? linePositionWriteConverter = null;
+                if (document is not null)
+                {
+                    linePositionWriteConverter = new();
+                    writeOptions.Converters.Add(linePositionWriteConverter);
+                }
+
+                // Serialize the TResponse and return it to the extension.
+                var responseType = resultProperty.PropertyType;
+                var responseJson = JsonSerializer.Serialize(result, responseType, writeOptions);
+
+                return new HandleCustomMessageResponse()
+                {
+                    Response = responseJson,
+                    Positions = linePositionWriteConverter?.LinePositions.OrderBy(lp => lp.Value).Select(lp => lp.Key).ToImmutableArray()
+                        ?? ImmutableArray<LinePosition>.Empty,
+                };
+            }
+            catch (Exception e)
+            {
+                Log(TraceEventType.Error, $"Custom handler {assemblyPath} {typeFullName} error: {e}");
+                throw;
+            }
+        }, cancellationToken);
+#endif
+    }
+
+#pragma warning disable CA1822 // Mark members as static
+    public ValueTask UnloadCustomMessageHandlerAsync(
+        string assemblyPath,
+        CancellationToken cancellationToken)
+#pragma warning restore CA1822 // Mark members as static
+    {
+#if NETSTANDARD2_0
+        throw new InvalidOperationException("Custom handlers are only supported in the servicehub host");
+#else
+
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        Requires.NotNullOrEmpty(assemblyPath);
+
+        return RunServiceAsync((_) =>
+        {
+            try
+            {
+                var defaultLoadContext = System.Runtime.Loader.AssemblyLoadContext.GetLoadContext(typeof(RemoteCustomMessageHandlerService).Assembly);
+                Contract.ThrowIfNull(defaultLoadContext);
+
+                if (_extensions.TryRemove(assemblyPath, out var extension))
+                {
+                    extension.Handlers.Clear();
+                    extension.AssemblyLoadContext.Unload();
```

Comment:
Consider implementing `Dispose` on `CustomMessageHandlerExtension` which does this